consensus/hotstuff/safetyrules/safety_rules_test.go

```diff
@@ -1,165 +1,423 @@
-}
+	expectedSafetyData := &hotstuff.SafetyData{
+		LockedOneChainView:      s.proposal.Block.QC.View,
+		HighestAcknowledgedView: s.proposal.Block.View,
+	}
 
-func testBelowVote(t *testing.T) {
-	// curView < blockView
-	blockView, curView, lastVotedView, isBlockSafe, isCommitteeMember := uint64(3), uint64(2), uint64(2), true, true
+	expectedVote := makeVote(s.proposal.Block)
+	s.signer.On("CreateVote", s.proposal.Block).Return(expectedVote, nil).Once()
+	s.persister.On("PutSafetyData", expectedSafetyData).Return(nil).Once()
 
-	// create voter
-	block, _, voter := createVoter(blockView, lastVotedView, isBlockSafe, isCommitteeMember)
+	vote, err := s.safety.ProduceVote(s.proposal, s.proposal.Block.View)
+	require.NoError(s.T(), err)
+	require.NotNil(s.T(), vote)
+	require.Equal(s.T(), expectedVote, vote)
+}
 
-	_, err := voter.ProduceVote(block, curView)
-	require.Error(t, err)
-	require.Contains(t, err.Error(), "expecting block for current view")
-	require.False(t, model.IsNoVoteError(err))
+// TestProduceVote_InvalidCurrentView tests that no vote is created if proposal is for invalid view
+func (s *SafetyRulesTestSuite) TestProduceVote_InvalidCurrentView() {
+	vote, err := s.safety.ProduceVote(s.proposal, s.proposal.Block.View+1)
+	require.Nil(s.T(), vote)
+	require.Error(s.T(), err)
+	s.persister.AssertNotCalled(s.T(), "PutSafetyData")
 }
 
-func testAboveVote(t *testing.T) {
-	// curView > blockView
-	blockView, curView, lastVotedView, isBlockSafe, isCommitteeMember := uint64(3), uint64(4), uint64(2), true, true
+// TestProduceVote_NodeEjected tests that no vote is created if block proposer is ejected
+func (s *SafetyRulesTestSuite) TestProduceVote_ProposerEjected() {
+	*s.committee = mocks.DynamicCommittee{}
+	s.committee.On("IdentityByBlock", s.proposal.Block.BlockID, s.proposal.Block.ProposerID).Return(nil, model.NewInvalidSignerErrorf("node-ejected")).Once()
+
+	vote, err := s.safety.ProduceVote(s.proposal, s.proposal.Block.View)
+	require.Nil(s.T(), vote)
+	require.True(s.T(), model.IsNoVoteError(err))
+	s.persister.AssertNotCalled(s.T(), "PutSafetyData")
+}
 
-	// create voter
-	block, _, voter := createVoter(blockView, lastVotedView, isBlockSafe, isCommitteeMember)
+// TestProduceVote_InvalidProposerIdentity tests that no vote is created if there was an exception retrieving proposer identity
+func (s *SafetyRulesTestSuite) TestProduceVote_InvalidProposerIdentity() {
+	*s.committee = mocks.DynamicCommittee{}
+	exception := errors.New("invalid-signer-identity")
+	s.committee.On("IdentityByBlock", s.proposal.Block.BlockID, s.proposal.Block.ProposerID).Return(nil, exception).Once()
 
-	_, err := voter.ProduceVote(block, curView)
-	require.Error(t, err)
-	require.Contains(t, err.Error(), "expecting block for current view")
-	require.False(t, model.IsNoVoteError(err))
+	vote, err := s.safety.ProduceVote(s.proposal, s.proposal.Block.View)
+	require.Nil(s.T(), vote)
+	require.ErrorAs(s.T(), err, &exception)
```

Comment:
Well, both `As` and `Is` are checking if this error is in the chain I believe, don't they?

Reply:
They both check the chain of errors, but `As` is checking type equality (any error from `errors` package would match, including some error instance other than `exception`) whereas `Is` is checking instance equality (only the specific `exception` error created in the test case would match). Here's an example: https://go.dev/play/p/Malz3OGjaoo

Reply:
Thanks for example but as I understand it's still good in our case since we use `errors` package only for exceptions, and point of those test is to find out if the exception or sentinel was returned. `errors` can't be cast to our custom sentinels(they don't implement `As`). Isn't that right? 
I've changed everything to `ErrorIs`

Reply:
Thanks for updating, I'll try to better explain the reasoning 👇 

> point of those test is to find out if the exception or sentinel was returned

We wouldn't accidentally confuse a sentinel for an exception, this is more the case I'm thinking of:

```go
func ProduceVote() {
  // a component we do not mock failing can cause an exception to be returned
  // we don't expect this error as part of this hypothetical test case
  ok := component1.CheckSomething()
  if !ok {
    return errors.New("some exception")
  }

  // we mock component2 to return an exception that we expect to be returned
  err := component2.DoSomething()
  if err != nil {
    return err
  }
  // ...
}
```
If we use `As`, we're not sure whether the error returned is from `component2` (expected) or `component1` (unexpected). If we use `Is`, we can be sure exactly the error we expected (from `component2`) was returned. 

It's more a general good practice than a concrete problem with the test as it exists currently.

Reply:
Thanks, that makes sense!

Reply:
The problem with `errors.Is` is that once an error is unwrap the error, but `errors.As` is able to.  See example
https://go.dev/play/p/95CRSMECAep